src/rspq/rspq.c

```diff
@@ -0,0 +1,1235 @@
+     * feature to raise RSP interrupts, while waiting for the previous
+     * interrupt to be processed (coalescing interrupts would cause syncpoints
+     * to be missed).
+     */
+    RSPQ_CMD_TEST_WRITE_STATUS = 0x08
+};
+
+
+// Make sure that RSPQ_CMD_WRITE_STATUS and RSPQ_CMD_TEST_WRITE_STATUS have
+// an even ID number. This is a small trick used to save one opcode in
+// rsp_queue.S (see cmd_write_status there for an explanation).
+/// @cond
+_Static_assert((RSPQ_CMD_WRITE_STATUS & 1) == 0);
+_Static_assert((RSPQ_CMD_TEST_WRITE_STATUS & 1) == 0);
+/// @endcond
+
+/** @brief Smaller version of rspq_write that writes to an arbitrary pointer */
+#define rspq_append1(ptr, cmd, arg1) ({ \
+    ((volatile uint32_t*)(ptr))[0] = ((cmd)<<24) | (arg1); \
+    ptr += 1; \
+})
+
+/** @brief Smaller version of rspq_write that writes to an arbitrary pointer */
+#define rspq_append2(ptr, cmd, arg1, arg2) ({ \
+    ((volatile uint32_t*)(ptr))[1] = (arg2); \
+    ((volatile uint32_t*)(ptr))[0] = ((cmd)<<24) | (arg1); \
+    ptr += 2; \
+})
+
+/** @brief Smaller version of rspq_write that writes to an arbitrary pointer */
+#define rspq_append3(ptr, cmd, arg1, arg2, arg3) ({ \
+    ((volatile uint32_t*)(ptr))[1] = (arg2); \
+    ((volatile uint32_t*)(ptr))[2] = (arg3); \
+    ((volatile uint32_t*)(ptr))[0] = ((cmd)<<24) | (arg1); \
+    ptr += 3; \
+})
+
+/** @brief Write an internal command to the RSP queue */
+#define rspq_int_write(cmd_id, ...) rspq_write(0, cmd_id, ##__VA_ARGS__)
+
+static void rspq_crash_handler(rsp_snapshot_t *state);
+static void rspq_assert_handler(rsp_snapshot_t *state, uint16_t assert_code);
+
+/** The RSPQ ucode */
+DEFINE_RSP_UCODE(rsp_queue, 
+    .crash_handler = rspq_crash_handler,
+    .assert_handler = rspq_assert_handler);
+
+/** 
+ * @brief The header of the overlay in DMEM.
+ * 
+ * This structure is placed at the start of the overlay in DMEM, via the
+ * RSPQ_OverlayHeader macros (defined in rsp_queue.inc).
+ */
+typedef struct rspq_overlay_header_t {
+    uint16_t state_start;       ///< Start of the portion of DMEM used as "state"
+    uint16_t state_size;        ///< Size of the portion of DMEM used as "state"
+    uint16_t command_base;      ///< Primary overlay ID used for this overlay
+    uint16_t reserved;          ///< Unused
+    uint16_t commands[];
+} rspq_overlay_header_t;
+
+/** @brief A pre-built block of commands */
+typedef struct rspq_block_s {
+    uint32_t nesting_level;     ///< Nesting level of the block
+    uint32_t cmds[];            ///< Block contents (commands)
+} rspq_block_t;
+
+/** @brief A RSPQ overlay ucode. This is similar to rsp_ucode_t, but is used
+ * internally to managed it as a RSPQ overlay */
+typedef struct rspq_overlay_t {
+    uint32_t code;              ///< Address of the overlay code in RDRAM
+    uint32_t data;              ///< Address of the overlay data in RDRAM
+    uint32_t state;             ///< Address of the overlay state in RDRAM (within data)
```

Comment:
Yes, the CPU knows the state size, but it's not relevant for this struct. Since the saved state is always located within the overlay data (it should be anyway, I'll add some asserts for this), the RSP will load both in a single DMA, so it doesn't need to know the state size for this step. When the overlay is unloaded however (or rather a new overlay is loaded), it does need the size for saving the state back to RDRAM. But because the overlay is loaded at that point, it can get the state size from the header. See rsp_queue.inc, lines 326-328.

Some basic checking of the state size is done in `rspq_overlay_register_internal` (lines 832-834). But as I said, it wouldn't hurt to also check that the state is contained within the overlay data.